src/wasm/wasm-validator.cpp

```diff
@@ -2437,12 +2438,41 @@ void FunctionValidator::visitTableCopy(TableCopy* curr) {
                     curr,
                     "table.copy source must have right type for dest");
   }
+  shouldBeEqualOrFirstIsUnreachable(curr->dest->type,
+                                    destTable->indexType,
+                                    curr,
+                                    "table.copy dest must be valid");
+  shouldBeEqualOrFirstIsUnreachable(curr->source->type,
+                                    sourceTable->indexType,
+                                    curr,
+                                    "table.copy source must be valid");
+  shouldBeEqualOrFirstIsUnreachable(curr->size->type,
+                                    destTable->indexType,
+                                    curr,
+                                    "table.copy size must be valid");
```

Comment:
This actually needs to be i32 if either of the input tables are 32-bit and i64 otherwise.